src/python/pants/util/objects.py

```diff
@@ -330,37 +310,79 @@ def validate_satisfied_by(self, obj):
       "value {!r} (with type {!r}) must satisfy this type constraint: {!r}."
       .format(obj, type(obj).__name__, self))
 
-  def __hash__(self):
-    return hash((type(self), self._types))
-
-  def __eq__(self, other):
-    return type(self) == type(other) and self._types == other._types
-
   def __ne__(self, other):
     return not (self == other)
 
   def __str__(self):
-    if self._desc:
-      constrained_type = '({})'.format(self._desc)
+    return '{}{}'.format(self._variance_symbol, self._description)
+
+
+class BasicTypeConstraint(TypeConstraint):
+  """A `TypeConstraint` predicated only on the object's type."""
+
+  # TODO: make an @abstract_classproperty decorator to do this boilerplate!
+  @classproperty
+  def _variance_symbol(cls):
+    """This is propagated to the the `TypeConstraint` constructor."""
+    raise NotImplementedError('{} must implement the _variance_symbol classproperty!'
+                              .format(cls.__name__))
+
+  def __init__(self, *types):
+    """Creates a type constraint centered around the given types.
+
+    The type constraint is satisfied as a whole if satisfied for at least one of the given types.
+
+    :param type *types: The focus of this type constraint.
```

Comment:
What does "focus" mean?

Reply:
It was copied from the `TypeConstraint` docstring -- I will expand this!